include/seastar/core/semaphore.hh

```diff
@@ -249,6 +249,26 @@ public:
     {
         static_assert(std::is_nothrow_move_constructible_v<expiry_handler>);
     }
+    basic_semaphore(basic_semaphore&& x) noexcept(std::is_nothrow_move_constructible_v<exception_factory>)
+        : exception_factory(std::move(x))
+        , _count(x._count)
+        , _wait_list(expiry_handler{*this})
+    {
+        // semaphore cannot be moved with non-empty waiting list
+        assert(x._wait_list.empty());
+    }
+
+    basic_semaphore& operator=(basic_semaphore&& x) noexcept(std::is_nothrow_move_assignable_v<exception_factory>) {
+        // semaphore cannot be moved with non-empty waiting list
+        assert(_wait_list.empty());
+        assert(x._wait_list.empty());
+        if (this != &x) {
+            exception_factory::operator=(x);
+            _count = x._count;
```

Comment:
Maybe I'm missing something here, but don't you have to reset _wait_list? If you don't, then why insist that it's empty?

Reply:
this->_wait_list remains intact, ready for business for new waiters, therefore being empty is good enough that we can toss away the other.wait_list.

Reply:
I understood why you want to check that x._wait_list.empty() because it' being tossed away, but why assert that this->wait_list is empty as well? Why do we care if there are waiters already, since the pointer to this remains value?

Reply:
I'd like to avoid assigning to a used semaphore. That's even broader than non-empty wait list as it means there could be outstanding semaphore units that relate to the pre-assigned state of the semaphore and when returned, would increment the count that may now be changed as a side effect of the assignment.
I think we better restrict the usage to a safe model that is easy to remember than to deal with weird corner cases even if we can technically support that.

Reply:
I'd like to avoid assigning to a used semaphore. That's even broader than non-empty wait list as it means there could be outstanding semaphore units that relate to the pre-assigned state of the semaphore and when returned, would increment the count that may now be changed as a side effect of the assignment.
I think we better restrict the usage to a safe model that is easy to remember than to deal with weird corner cases even if we can technically support that.